From 2ac4e1a8c01fcc8840b942079e5ce8caf7172d0f Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Tue, 22 Nov 2016 16:48:50 +0800
Subject: [PATCH 1142/2241] usb: gadget: automatically select correct/working
 udc driver for configfs

- For Kernel USB Gadget Configfs interface, a UDC driver has to be selected
  in advance with command: echo "udc_driver_name" > UDC. It is not very
  convenient or reasonable for SoCs that have both USB2 and USB3 device
  controller, like Armada3700/38x/8K, since user would have to know which
  type of USB Host (USB2 or USB3) it will be connected in advance to bind
  the right udc driver. And if Host mode changed between USB2 and USB3,
  user has to unbind the outdated UDC driver first, then bind the new one.
- This patch introduces mvebu gadget glue layer driver to configfs, who:
  * ignores which udc driver it is in configfs command, and always
    load the correct Marvell USB2 and USB3 device driver.
  * automatically unload the outdated udc driver and load the new one
    when USB Host type changes between USB2 and USB3.

Change-Id: I2bc50087f23ec41b08ac2dd33576560cf7c8378c
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/33945
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
---
 drivers/usb/gadget/udc/udc-core.c | 9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/drivers/usb/gadget/udc/udc-core.c b/drivers/usb/gadget/udc/udc-core.c
index 302fbb5..5d29050 100644
--- a/drivers/usb/gadget/udc/udc-core.c
+++ b/drivers/usb/gadget/udc/udc-core.c
@@ -506,12 +506,9 @@ int usb_udc_attach_driver(const char *name, struct usb_gadget_driver *driver)
 	int ret = -ENODEV;
 
 	mutex_lock(&udc_lock);
-	list_for_each_entry(udc, &udc_list, list) {
-		ret = strcmp(name, dev_name(&udc->dev));
-		if (!ret)
-			break;
-	}
-	if (ret) {
+
+	udc = udc_detect(&udc_list, driver);
+	if (!udc) {
 		ret = -ENODEV;
 		goto out;
 	}
-- 
2.7.4

